[ILLink analyzer] Add analyzer support for feature checks

sbomer opened the pull request:
Adds support for annotating static boolean properties with `[FeatureCheckAttribute(typeof(RequiresDynamicCodeAttribute))]`, causing the property to be treated as a guard for analyzer warnings about the corresponding `Requires` attribute.

Adds two new warnings for:
- Invalid use of `FeatureCheckAttribute` applied to a non-static or non-bool property
- Implementation of the property doesn't obviously satisfy the "guard property" (it should return false whenever the guarded feature is disabled).

See https://github.com/dotnet/runtime/pull/94625 for notes on the design.
See https://github.com/dotnet/runtime/issues/96859 for the API proposal.

sbomer committed:
Add analyzer support for feature guards

sbomer committed:
Merge remote-tracking branch 'origin/main' into featureGuardAnalyzerSupport

sbomer committed:
FeatureGuard -> FeatureCheck

sbomer committed:
Add support for FeatureDependsOn

sbomer committed:
Add more tests, tweak guard validation

sbomer committed:
Support guards that return false

sbomer committed:
Clean up

sbomer committed:
Add missing check, split up guard body tests

jtschuster commented on src/tools/illink/src/ILLink.RoslynAnalyzer/TrimAnalysis/TrimAnalysisReturnValuePattern.cs.cs:
```diff
@@ -0,0 +1,68 @@
+		{
+			var diagnosticContext = new DiagnosticContext (Operation.Syntax.GetLocation ());
+			// For now, feature check validation is enabled only when trim analysis is enabled.
+			if (context.EnableTrimAnalyzer) {
```
nit: maybe return early here to avoid the deep nesting.

jtschuster commented on src/tools/illink/src/ILLink.RoslynAnalyzer/TrimAnalysis/TrimAnalysisReturnValuePattern.cs.cs:
```diff
@@ -0,0 +1,68 @@
+					return diagnosticContext.Diagnostics;
+				}
+
+				if (ReturnValue != FeatureChecksValue.All) {
```
nit: return early here to avoid deep nesting

jtschuster commented on src/tools/illink/src/ILLink.RoslynAnalyzer/TrimAnalysis/TrimAnalysisReturnValuePattern.cs.cs:
```diff
@@ -0,0 +1,68 @@
+
+namespace ILLink.RoslynAnalyzer.TrimAnalysis
+{
+	public readonly record struct TrimAnalysisReturnValuePattern
```
What does this represent? Is it always the return value of a [FeatureCheck] property in a method, or [FeatureCheck] property getter definition?

jtschuster commented on src/tools/illink/test/Mono.Linker.Tests.Cases/DataFlow/FeatureCheckAttributeDataFlow.cs:
```diff
@@ -0,0 +1,615 @@
+			// No warning for this case because we don't validate that the attribute usage matches
+			// the expected AttributeUsage.Property for assemblies that define their own version
+			// of FeatureCheckAttributes.
+			[FeatureCheck (typeof(RequiresUnreferencedCodeAttribute))]
```
Should we add an issue link?

jtschuster reviewed (approved):
Thank you!

jtschuster commented on src/tools/illink/src/ILLink.RoslynAnalyzer/RequiresAnalyzerBase.cs:
```diff
@@ -19,9 +20,7 @@ public abstract class RequiresAnalyzerBase : DiagnosticAnalyzer
-		internal abstract string FeatureName { get; }
-
-		private protected abstract string RequiresAttributeFullyQualifiedName { get; }
+		internal abstract string RequiresAttributeFullyQualifiedName { get; }
```
Maybe it makes sense to keep the RequiresAttributeFullQualifiedName and FeatureTypeName separated so it's easier to translate for a generic FeatureSwitch analyzer?

sbomer committed:
Update src/tools/illink/src/ILLink.RoslynAnalyzer/DataFlow/FeatureChecksValue.cs

Co-authored-by: Jackson Schuster <36744439+jtschuster@users.noreply.github.com>

sbomer committed:
Update src/tools/illink/src/ILLink.Shared/SharedStrings.resx

Co-authored-by: Jackson Schuster <36744439+jtschuster@users.noreply.github.com>

sbomer committed:
PR feedback

- Add more detail to feature check validation message
- Use existing attribute helper to get AttributeData by attribute name
- Replace unnecessary ImmutableArray Builder
- Invert some if conditions to return early
- Rename TrimAnalysisReturnValuePattern -> FeatureCheckReturnValuePattern
